src/container-mapping.ts

```diff
@@ -81,71 +80,102 @@ export class ContainerMapping implements IMicrosoftSecurityDevOps {
+            });
+
+        if (!bearerToken) {
+            throw new Error("Empty OIDC token received");
+        }
+
+        var reportSent: boolean = await this.sendReport(JSON.stringify(reportData), bearerToken, sendReportRetryCount);
+        if (!reportSent) {
+            throw new Error("Unable to send report to backend service");
+        };
+    }
+
+    /**
+     * Execute command and setup the listener to capture the output
+     * @param command Command to execute
+     * @param listener Listener to capture the output
+     * @returns a Promise
+     */
+    private async execCommand(command: string, listener: string[]): Promise<void> {
+        return exec.exec(command, null, {
             listeners: {
                 stdout: (data: Buffer) => {
-                    reportData.dockerImages = reportData.dockerImages.concat(data.toString());
+                    var lines = data.toString().trim().split(os.EOL);
+                    lines.forEach(element => {
+                        listener.push(element);
+                    });
                 }
             }
+        })
+        .then((exitCode) => {
+            if (exitCode != 0) {
+                Promise.reject(`Command failed with exit code ${exitCode}`);
+            }
         });
-
-        core.debug("Finished data collection, starting API calls.");
-
-        await this.sendReport(reportData, sendReportRetryCount);
     }
 
     /**
      * Sends a report to Defender for DevOps and retries on the specified count
      * @param data the data to send
      * @param retryCount the number of time to retry
-     * @returns a Promise
+     * @returns a boolean Promise to indicate if the report was sent successfully or not
      */
-    private async sendReport(data: Object, retryCount: number = 0): Promise<void> {
-        return new Promise(async (resolve, reject) => {
-            do {
-                try {
-                    await this._sendReport(data);
-                    resolve();
-                    break;
-                } catch (error) {
-                    if (retryCount == 0) {
-                        reject('Failed to send report: ' + error);
-                    } else {
-                        retryCount--;
-                        core.debug(`Retrying API call. Retry count: ${retryCount}`);
-                    }
+    private async sendReport(data: string, bearerToken: string, retryCount: number = 0): Promise<boolean> {
+        core.debug(`attempting to send report: ${data}`);
+        return await this._sendReport(data)
+            .then(() => {
+                return true;
+            })
+            .catch((error) => {
+                if (retryCount == 0) {
+                    return false;
+                } else {
+                    core.warning(`Retrying API call due to error: ${error}.\nRetry count: ${retryCount}`);
+                    retryCount--;
+                    return this.sendReport(data, bearerToken, retryCount);

```

Comment:
Switching to recursive, do you need to await this as well? or on line `158` is `await` in `return await this._sendReport(...)` not required? As the caller will wait.

Reply:
ohh good catch! I honestly didn't even notice it! 
The reason it was working was because the internal function was also not-async, causing the whole flow to be synchronous.

```js
(error) => {
                if (retryCount == 0) {
                    return false;
                } else {
                    writeToOutStream(`Retrying API call due to error: ${error}.\nRetry count: ${retryCount}`);
                    retryCount--;
                    return this.sendReport(data, bearerToken, retryCount);
                }
            }
```

 I've made it all async to keep it consistent and easier to understand.